src/backend/sham/generate-sham.rkt

```diff
@@ -0,0 +1,520 @@
+         [else ;; evaluate values for their effects
+          (block (map generate-effect exp*))])]
+      [(or
+        (Stack-Alloc _)
+        (Var _)
+        (Global _)
+        (Code-Label _)
+        (Quote _)
+        (No-Op))
+       (svoid)]
+      [(Switch v (list (cons (list lhs*) rhs*) ...) default)
+       (switch
+        v
+        (for/list ([n lhs*]) (csi n i64))
+        (map generate-effect rhs*)
+        (generate-effect default))]))
+
+  (define alloc
+    (call
+     (re #f
+         (case (garbage-collector)
+           [(Boehm) 'GC_malloc]
+           [(None) 'nonegc_malloc])
+         i64)))
+  
+  (define (generate-op p exp*)
+    (match* (p exp*)
+      [('Array-set! (list a o v))
+       (store! v (gep (->arr-ptr a) (list o)))]
+      [('Array-ref (list a o))
+       (load (gep (->arr-ptr a) (list o)))]
+      [('print-float (list f p))
+       (app^ (re #f 'printf tvoid) (cstring "%.*.lf") p (->float f))]
+      [('print-int (list d))
+       (app^ (re #f 'printf tvoid) (cstring "%ld") d)]
+      [('print-bool (list b))
+       (if^ (imdt->bool b)
+            (app^ #f (re #f 'puts tvoid) (cstring "#t"))
+            (app^ #f (re #f 'puts tvoid) (cstring "#f")))]
+      [('Printf (cons fmt exp*))
+       (app (re #f 'printf tvoid) (cons fmt exp*))]
+      [('Alloc (list exp)) (alloc (mul size-of-imdt exp))]
+      [('Types-hashcons! (and args (list e hcode)))
+       (app (re #f 'hashcons i64) (cons types_ht args))]
+      [('Types-gen-index! (list))
+       (store! types_unique_index_counter (add types_unique_index_counter (cui 1 i64)))]
+      [('mref-cast-queue-enqueue (and args (list addr ty)))
+       (app (re #f 'cast_queue_enqueue tvoid) (cons mref-cast-q args))]
+      [('mref-cast-queue-dequeue (list))
+       (app (re #f 'cast_queue_dequeue) (list mref-cast-q))]
+      [('mref-cast-queue-not-empty? (list))
+       (app (re #f 'cast_queue_is_not_empty i64) (list mref-cast-q))]
+      [('mref-cast-queue-peek-address (list))
+       (app (re #f 'cast_queue_peek_address i64) (list mref-cast-q))]
+      [('mref-cast-queue-peek-type (list))
+       (app (re #f 'cast_queue_peek_type i64) (list mref-cast-q))]
+      [('mvect-cast-queue-enqueue (and args (list addr ty)))
+       (app (re #f 'cast_queue_enqueue tvoid) (cons mvect-cast-q args))]
+      [('mvect-cast-queue-dequeue (list))
+       (app (re #f 'cast_queue_dequeue) (list mvect-cast-q))]
+      [('mvect-cast-queue-not-empty? (list))
+       (app (re #f 'cast_queue_is_not_empty i64) (list mvect-cast-q))]
+      [('mvect-cast-queue-peek-address (list))
+       (app (re #f 'cast_queue_peek_address i64) (list mvect-cast-q))]
+      [('mvect-cast-queue-peek-type (list))
+       (app (re #f 'cast_queue_peek_type i64) (list mvect-cast-q))]
+      ;; [('timer-start (list)) (display timer-start)]
+      ;; [('timer-stop  (list)) (display timer-stop)]
+      ;; [('timer-report (list)) (display timer-report)]
```

Comment:
You added those primitives to the `entry.c` file, so you probably want to delete these lines here.